### ISSUE TITLE
Delete a word from the dictionary?

### ISSUE THREAD

**Issue by keforbes** [florisboard/florisboard#737]
When using glide typing, if I try gliding the word `the` I constantly get `thr`.  Now, I'll admit that my gliding probably stopped short of the `e` and ended up on the `r` so it isn't _wrong_ necessarily, but I don't think `thr` is even a word so there will never be a scenario where I'm intending to type `thr`.  I've had a couple other cases where my attempt at gliding a word comes up with some random letters that I wouldn't have thought was a word at all.  Of course, in every scenario, if I was more precise with my gliding it would've picked the correct word, so this isn't a defect by any means.  I'm just wondering if there's any way for me to delete words from the dictionary that I will never use.

**Comment by Glitchy-Tozier** [comment]
Currently, there's no way to do this in-app. It'll probably be added eventually though.

If you're really interested in going deep and editing files to accomplish this behavior, you can probably ask `@patrickgold` for help.

**Comment by Glitchy-Tozier** [comment]
Actually, in the "writing"-section, there are two options to "manage user dictionary".

Does that work for you?

**Comment by keforbes** [comment]
Nope, both those sections are empty for me because I haven't added any custom words.  These are words that seem to be in florisboard's dictionary.

**Comment by linareyne** [comment]
This would be great, in addition to learning user's common words, a bit of context awareness, and allowing the glide paths to be imperfect. So frustrating getting stuff like `yoy` instead of `you` when gliding. The english dictionary has a ton of nonsense/non-english words, surnames, random letters with diacritic marks, etc. And the glide path has to pretty much be perfectly precise to even get the word you wanted to appear in the suggestions (assuming you have >3 suggestions enabled, since often the first few are shit like `ordd` or `û`).

Attempting `do` (I can see `so` being the triggered word, but none of the others are even english, come on):
![signal-2021-06-21-121743](https://user-images.githubusercontent.com/20880695/122795001-da9d6500-d28a-11eb-8924-ced751d2378c.jpeg)

Attempting `you` (I mean, wtf):
![signal-2021-06-21-121754](https://user-images.githubusercontent.com/20880695/122795005-dbce9200-d28a-11eb-90c6-a8f2e2fa3022.jpeg)

Attempting `today`:
![signal-2021-06-21-121803](https://user-images.githubusercontent.com/20880695/122795009-dc672880-d28a-11eb-8a42-aec0bb1b443b.jpeg)

I can successfully glide those words a hundred times without looking on other keyboards, but going slow and staring at what I'm doing fails regularly here). Having used a number of swipe/glide keyboards, I've honestly never come across anything remotely close to this bad. This is an exciting project, so I'm def sticking around, but sadly the keyboard is nearly unusable for me at the moment because of this.

**Comment by keforbes** [comment]
Thank you!  I keep getting `yoy` also.  I'm glad to know it isn't just me.

**Comment by williamtheaker** [comment]
I'd rather have the English dictionary cleaned as part of the import process to remove unwanted strings like thr, dont't, and yoy, instead of manual curation by each user.

**Comment by Gitoffthelawn** [comment]
This is probably my primary request for this app at this time.  I apparently write the word "you" quite a bit, and I keep getting "yoy".

I'm half-tempted to write everyone I know and just explain that when I write "yoy", I mean "you". :)  But I really, really don't want to read all the "just get an iPhone" or "just use Google's keyboard" replies. No, and no. :)

**Comment by Glitchy-Tozier** [comment]
> I apparently write the word "you" quite a bit, and I keep getting "yoy".

I understand this is annoying.

One thing that could fix this issue, as you have suggested, is adding the option to delete the word. Another thing should be to improve the algorithm with which words are chosen.

As far as I'm aware, both those things are part of the next big goal: https://github.com/florisboard/florisboard/blob/master/ROADMAP.md#florisboards-feature-roadmap--milestones

**Comment by Gitoffthelawn** [comment]
> > I apparently write the word "you" quite a bit, and I keep getting "yoy".
> 
> I understand this is annoying.
> 
> One thing that could fix this issue, as you have suggested, is adding the option to delete the word. Another thing should be to improve the algorithm with which words are chosen.
> 
> As far as I'm aware, both those things are part of the next big goal: https://github.com/florisboard/florisboard/blob/master/ROADMAP.md#florisboards-feature-roadmap--milestones

Thanks for your constructive feedback.  I hope I can favour you with a reply in a similar helpful sprit.

I think improving suggestions is always a good idea.  Except with perhaps the very best suggestion engine (and the very best requires a fair amount of dealing with the semantics of human written languages; I worked on human language translation tools years ago, so I know this is no small task), I hypothesize the option to delete (and add) words will still be needed.

As a humorous example, let's say one user often types the word "bump", but never types the word "hump" ("b" and "h" are diagonally adjacent in QWERTY).  Removing the word "hump" from their dictionary (but hopefully not their life!) may improve their satisfaction.

Adding words to the dictionary may be valuable as well.  For example, in text messages, I commonly want to write "ya" (instead of "yes"), which isn't a word in most dictionaries.  On the other hand, I rarely write "ha" ("h" and "y" are diagonally adjacent in QWERTY), so I would prefer to add "ya" and remove "ha" (or at least specify a priority).

Ideally, the keyboard will (optionally) keep track of how often the user uses each word, and make recommendations accordingly (and adding semantic knowledge of the written language will be a big plus too).  This, of course, can be a privacy concern.  One way around this privacy issue is to only keep track of the use of common words (and of course to never sync any such data with a cloud server).

I hope this is helpful.

I'm quite busy, but I have a reasonable amount of experience (and formal education) in this area, so feel free to pick my brain.  I'll do my best to lend a hand (even though auto-suggest may recommend that I'm willing to lend a *band*).

**Comment by Glitchy-Tozier** [comment]
> This, of course, can be a privacy concern. One way around this privacy issue is to only keep track of the use of common words (and of course to never sync any such data with a cloud server).

One neat thing is that Patrick – the dev of FlorisBoard – is dedicated to not introduce any internet-permissions. Additionally, I can remember discussions about how we could preserve user privacy, but I can't remember where they took place.

> I'm quite busy, but I have a reasonable amount of experience (and formal education) in this area, so feel free to pick my brain. I'll do my best to lend a hand (even though auto-suggest may recommend that I'm willing to lend a band).

Thank you for your willingness to bend a land, let's see what Patrick will do! :)

**Comment by Glitchy-Tozier** [comment]
@ghtesting2020 Currently, the main developer of FlorisBoard is a bit occupied by other parts of life, which is why there isn't much progress at the moment.
However, I think he seriously enjoys working on Florisboard and will return to it eventually.

As I understand it, the main difficulty with improving glide-typing is that it requires improvements to word suggestions, which are the next planned feature, but are pretty complex to implement.
